rdflib/graph.py

```diff
@@ -670,35 +670,120 @@ def set(self, triple):
-        for s, p, o in self.triples((subject, None, object)):
-            yield p
-
-    def objects(self, subject=None, predicate=None) -> Iterable[Node]:
-        """A generator of objects with the given subject and predicate"""
-        for s, p, o in self.triples((subject, predicate, None)):
-            yield o
-
-    def subject_predicates(self, object=None):
+    def subjects(self, predicate=None, object=None, unique=False) -> Iterable[Node]:
+        """A generator of unique subjects with the given predicate and object"""
+        if not unique:
+            for s, p, o in self.triples((None, predicate, object)):
+                yield s
+        else:
+            subs = set()
+            for s, p, o in self.triples((None, predicate, object)):
+                if s not in subs:
+                    yield s
+                    try:
+                        subs.add(s)
+                    except MemoryError as e:
+                        logger.error(
+                            f"{e}. Consider not setting parameter 'unique' to True"
+                        )
+                        raise
+
+    def predicates(self, subject=None, object=None, unique=False) -> Iterable[Node]:
+        """A generator of unique predicates with the given subject and object"""
+        if not unique:
+            for s, p, o in self.triples((subject, None, object)):
+                yield p
+        else:
+            preds = set()
+            for s, p, o in self.triples((subject, None, object)):
+                if p not in preds:
+                    yield p
+                    try:
+                        preds.add(p)
+                    except MemoryError as e:
+                        logger.error(
+                            f"{e}. Consider not setting parameter 'unique' to True"
+                        )
+                        raise
+
+    def objects(self, subject=None, predicate=None, unique=False) -> Iterable[Node]:
+        """A generator of unique objects with the given subject and predicate"""
+        if not unique:
+            for s, p, o in self.triples((subject, predicate, None)):
+                yield o
+        else:
+            objs = set()
+            for s, p, o in self.triples((subject, predicate, None)):
+                if o not in objs:
+                    yield o
+                    try:
+                        objs.add(o)
+                    except MemoryError as e:
+                        logger.error(
+                            f"{e}. Consider not setting parameter 'unique' to True"
+                        )
+                        raise
+
+    def subject_predicates(
+        self, object=None, unique=False
+    ) -> Generator[Tuple[Node, Node], None, None]:
         """A generator of (subject, predicate) tuples for the given object"""
-        for s, p, o in self.triples((None, None, object)):
-            yield s, p
+        if not unique or not hasattr(self.store, "identifier"):
```

Comment:
I'm not sure why I did not spot this before, but I'm unsure why this condition (and others) has `or not hasattr(self.store, "identifier")`. As far as I can tell this should work fine regardless of whether or not the store has an `identifier` - but maybe I'm missing something.

Reply:
And also if there is some reason why this cannot return unique triples for stores with identifiers then maybe it should rather throw a `TypeError` or `ValueError` as the caller will be expecting unique values.

Reply:
```suggestion
        if not unique:
```

Reply:
> And also if there is some reason why this cannot return unique triples for stores with identifiers then maybe it should rather throw a `TypeError` or `ValueError` as the caller will be expecting unique values.

For context, I initially trial all my PRs in the id-as-context venv in an attempt to keep that side of things coherent (It's demanding) and I think these specific checks must be related to errors arising from executing in the id-as-context venv. 

When I split off this change into its own branch, I removed the checks from the rebased PR code but omitted to remove them from the code I added. Thanks for the catch. I've removed the checks pro tem, edited the docstrings to be more explicit (anticipating expanding them further during the docs overhaul) and neatened the tests.